src/pg_query_summary_truncate.c

```diff
@@ -0,0 +1,673 @@
+				add_truncation_where_clause(state, node, stmt->whereClause);
+				break;
+			}
+
+		case T_RuleStmt:
+			{
+				RuleStmt *stmt = castNode(RuleStmt, node);
+				add_truncation_where_clause(state, node, stmt->whereClause);
+				break;
+			}
+
+		case T_CommonTableExpr:
+			{
+				CommonTableExpr *stmt = castNode(CommonTableExpr, node);
+
+				if (stmt->ctequery != NULL) {
+					PgQueryDeparseResult result = pg_query_deparse(node);
+
+					if (result.error) {
+						state->error = result.error;
+						return false;
+					}
+
+					int32_t length = strlen(result.query);
+
+					add_truncation(state,
+							TRUNCATION_CTE_QUERY,
+							node,
+							length);
+				}
+
+				break;
+			}
+
+		case T_InferClause:
+			{
+				InferClause *stmt = castNode(InferClause, node);
+				add_truncation_where_clause(state, node, stmt->whereClause);
+				break;
+			}
+
+		case T_OnConflictClause:
+			{
+				OnConflictClause *stmt = castNode(OnConflictClause, node);
+
+				if (stmt->targetList != NULL)
+					add_truncation(state,
+							TRUNCATION_TARGET_LIST,
+							node,
+							update_target_list_len(stmt->targetList));
+
+				add_truncation_where_clause(state, node, stmt->whereClause);
+
+				break;
+			}
+
+		default:
+			break;
+	}
+
+	if (!pg_query_raw_tree_walker_supports(node))
+		return false;
+
+	state->depth++;
+
+	return raw_expression_tree_walker(node, summary_truncation_options, (void *) state);
+}
+
+static int cmp_possible_truncation_depth(const ListCell *a, const ListCell *b)
+{
+	const PossibleTruncation *pt_a = lfirst(a);
+	const PossibleTruncation *pt_b = lfirst(b);
+
+	return pt_a->depth - pt_b->depth;
+}
+
+static void expand_ellipses(char *str)
+{
+	// https://www.compart.com/en/unicode/U+2026
+	// pattern is a double quote, ellipses, and another double quote.
+	char pattern[5] = {'"', 0xE2, 0x80, 0xA6, '"'};
+
+	for (size_t i = 0; i < strlen(str); i++) {
+		if (memcmp(str + i, pattern, 5) == 0) {
+			size_t len = strlen(str);
+			memcpy(str + i, "...", 3);
+			memmove(str + i + 3, str + i + 5, len - i - 2);
+		}
+	}
+}
+
+// This is equivalent to a kludge in the Rust implementation.
+// Not sure why the `SELECT ... AS ...` shows up in both Rust and C,
+// but not the Ruby implementation.
+static void remove_unneeded_as(char *str)
+{
+	char *pattern = "SELECT ... AS ...";
+	size_t plen = strlen(pattern);
+
+	size_t slen = strlen("SELECT ...");
+
+	for (size_t i = 0; i < strlen(str); i++) {
+		if (memcmp(str + i, pattern, plen) == 0) {
+			strcpy(str + i + slen, str + i + plen);
+		}
+	}
+}
+
+static PgQueryError *apply_truncations(Summary *summary, Node *tree, TruncationState *state, int truncation_limit)
+{
+	List *truncations = state->truncations;
+
+	PgQueryDeparseResult result = pg_query_deparse(tree);
+	char *output = result.query;
+
+	if (result.error)
+		return result.error;
+
+	if (strlen(result.query) <= truncation_limit) {
+		summary->truncated_query = output;
+		return NULL;
+	}
```

Comment:
We could do this before generating truncations to save on the generation if the query is already short enough.